### PR TITLE
add todo app on react ts

### PR DESCRIPTION
**Pull request by KaterynaMadzhara**

### CODE REVIEW

File: todo_react/todo-react/src/App.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import React, {useState} from 'react';
+import './App.css';
+import {AddTask} from "./components/AddTask";
+import {ToDoItem} from "./components/ToDoItem";
+import {ITodo} from "./models";
+
+export const App: React.FC = () => {
+    let todoList = [...JSON.parse(localStorage.getItem('TodoList') || "")]
+    const [todos, setTodos] = useState<Array<{ id: number, completed: boolean, title: string }>>(todoList)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: use type `ITodo[]` or `Array<ITodo>`

File: todo_react/todo-react/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    let emptyList = todoList.length === 0 ? "No Tasks added" : ""
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: 1. use `const` if variable is not reassigned 
2. variable `emptyList` means that we have some **List** (usually array `[]`). if you want to store some message/text call it `emptyListMessage` or even do teranry right in JSX (`<p className="empty">{todoList.length === 0 ? "No Tasks added" : ""}</p>`)

File: todo_react/todo-react/src/App.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const addTodo = (todo: ITodo) => {
+        setTodos([...todos, todo])
+        localStorage.setItem('TodoList', JSON.stringify([...todos, todo]))
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: I would move logic `localStorage.setItem('TodoList', JSON.stringify(todos))` to separate function:

```typescript
const saveToLocalStorage = (todos: ITodo[]) => {
   localStorage.setItem('TodoList', JSON.stringify(todos))
}
```

Comment (gerasimvol) [reply todo_react/todo-react/src/App.tsx]:
```typescript
const newTodos = [...todos, todo]
setTodos(newTodos)
saveToLocalStorage(newTodos)
```

File: todo_react/todo-react/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    const deleteTodo = (todo: ITodo) => {
+        const newTodos = todos.filter(todoEl => todo.id !== todoEl.id)
+        setTodos(newTodos)
+        localStorage.setItem('TodoList', JSON.stringify(newTodos))
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: use `saveToLocalStorage`

File: todo_react/todo-react/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    const doneTodo = (todo: ITodo) => {
+        const newTodos = todos.map(todoEl => {
+            if (todo.id === todoEl.id) {
+                todoEl.completed = !todoEl.completed
+            }
+            return todoEl
+        })
+        setTodos(newTodos)
+        localStorage.setItem('TodoList', JSON.stringify(newTodos))
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: use `saveToLocalStorage `

File: todo_react/todo-react/src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    const editTodo = (todo: ITodo, value: string) => {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: I would suggest passing no `todo`, but `todoId`

File: todo_react/todo-react/src/App.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        const newTodos = todos.map(todoEl => {
+            if (todo.id === todoEl.id) {
+                todoEl.title = value
+            }
+            return todoEl
+        })
+        setTodos(newTodos)
+        localStorage.setItem('TodoList', JSON.stringify(newTodos))
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: use `saveToLocalStorage`

File: todo_react/todo-react/src/components/ToDoItem.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+interface TodoProps {
+    todo: ITodo
+    deleteTodo: (todo: ITodo) => void
+    doneTodo: (todo: ITodo) => void
+    editTodo: (todo: ITodo, value: string) => void
+}
+
+export function ToDoItem(props: TodoProps) {
+    const [edit, setEdit] = useState(false)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/ToDoItem.tsx]: `edit` is a verb – used for functions (some action), we need `isEditing` – boolean value (status)

check something like https://github.com/kettanaito/naming-cheatsheet

File: todo_react/todo-react/src/components/ToDoItem.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    const [value, setValue] = useState<string>(props.todo.title)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/ToDoItem.tsx]: `value` seems too generic, maybe `currentTitle` or `editableTitle`

File: todo_react/todo-react/src/components/ToDoItem.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const editTitleInputRef = useRef<HTMLInputElement>(null)
+    const onClick = (value: string) => {
+        props.editTodo(props.todo, value)
+        setEdit(false)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/ToDoItem.tsx]: `setIsEditing` would be much clear

File: todo_react/todo-react/src/components/ToDoItem.tsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                               setValue(e.target.value)
+                           }}/>
+                    <button onClick={() => onClick(value)}>save</button>
+                </div>
+                :
+                <div className="task">
+                    <input className="text"
+                           style={{textDecoration: props.todo.completed ? 'line-through' : 'none'}}
+                           value={props.todo.title}
+                           readOnly/>
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/ToDoItem.tsx:38]: native html attr `readonly`

File: todo_react/todo-react/src/components/AddTask.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
+import React, {ChangeEvent, MouseEventHandler, useState} from 'react';
+import {ITodo} from "../models";
+
+interface AddTaskProps {
+    addTodo: (todo: ITodo) => void
+}
+
+export const AddTask: React.FC<AddTaskProps> = ({addTodo}) => {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/AddTask.tsx]: we operate with entities `Todo`, why `Task`?

File: todo_react/todo-react/src/App.tsx (hunk 1 above)
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: try to move `todos` and function that changes it to [React Context](https://reactjs.org/docs/context.html) for practice

File: todo_react/todo-react/src/components/ToDoItem.tsx (hunk 11 above)
Comment (KaterynaMadzhara) [reply todo_react/todo-react/src/components/ToDoItem.tsx:38]: didn't get what do you mean

File: todo_react/todo-react/src/index.tsx [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import './index.css';
+import App from './App';
+import reportWebVitals from './reportWebVitals';
+
+const root = ReactDOM.createRoot(
+  document.getElementById('root') as HTMLElement
+);
+root.render(
+  <React.StrictMode>
+    <App />
```
Comment (gerasimvol) [inline todo_react/todo-react/src/index.tsx]: wrap your `<App>` into `<TodosProvider>`

File: todo_react/todo-react/src/components/ToDoItem.tsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import {ITodo} from "../models";
+import {TodosContext} from "../context"
+interface TodoProps {
+    todo: ITodo
+}
+export function ToDoItem(props: TodoProps) {
+    const [isEditing, setIsEditing] = useState(false)
+    const [editableTitle, setEditableTitle] = useState<string>(props.todo.title)
+    const editTitleInputRef = useRef<HTMLInputElement>(null)
+    const todosContext = useContext(TodosContext)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/ToDoItem.tsx]: try archiving something like that
`const { editTodo, deleteTodo, doneTodo } = useTodosContext()`

File: todo_react/todo-react/src/App.tsx [hunk 15]
```diff
@@ -0,0 +1,60 @@
+import React, {useState} from 'react';
+import './App.css';
+import {AddToDo} from "./components/AddToDo";
+import {ToDoItem} from "./components/ToDoItem";
+import {ITodo} from "./models";
+import {TodosContext} from "./context"
+
+export const App: React.FC = () => {
+    let todoList = [...JSON.parse(localStorage.getItem('TodoList') || "")]
+    const [todos, setTodos] = useState<ITodo[]>(todoList)
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: `<TodosProvider>` should create `todos` as state, not `<App>`
`<App>` should only take `todos` from context via `useTodosContext` or `useContext(TodosContext)`

File: todo_react/todo-react/src/App.tsx [hunk 16]
```diff
@@ -0,0 +1,60 @@
+import React, {useState} from 'react';
+import './App.css';
+import {AddToDo} from "./components/AddToDo";
+import {ToDoItem} from "./components/ToDoItem";
+import {ITodo} from "./models";
+import {TodosContext} from "./context"
+
+export const App: React.FC = () => {
+    let todoList = [...JSON.parse(localStorage.getItem('TodoList') || "")]
```
Comment (Romms) [inline todo_react/todo-react/src/App.tsx]: 1. why need `...` ?
2. can try to clean local storage, fix JSON.parse error
3. avoid parsing data each render (read about react lifecycle)
4. Move state management to a new component
5. Re-render todos list when it's changed in local storage in a different tab

File: todo_react/todo-react/src/App.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    const deleteTodo = (todoId: number) => {
+        const newTodos = todos.filter(todoEl => todoId !== todoEl.id)
+        setTodos(newTodos)
+        saveToLocalStorage(newTodos)
+    }
+    const doneTodo = (todoId: number) => {
+        const newTodos = todos.map(todoEl => {
+            if (todoId === todoEl.id) {
+                todoEl.completed = !todoEl.completed
```
Comment (Romms) [inline todo_react/todo-react/src/App.tsx]: seems like it's a state modification, please avoid it :)

Comment (gerasimvol) [reply todo_react/todo-react/src/App.tsx]: Roman means you shouldn’t change `todoEl` (original object in `todos`). Create new object instead and return it from `.map` callback

File: todo_react/todo-react/src/App.tsx (hunk 16 above)
Comment (gerasimvol) [reply todo_react/todo-react/src/App.tsx]: some hints
1. `JSON.parse("[]")` returns array. you don't need `...`
2. wrap parsing in `try/catch`
3. you are making `JSON.parse` every time component re-renders, but you need it only once – initial state set (can be done on mounted)
4. move state logic (`useState` and `JSON.parse` to `<TodosProvider>`)
5. it is additional task (try something like `window.addEventListener("storage", () => {})`)

File: todo_react/todo-react/src/App.tsx [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            return todoEl
+        })
+        setTodos(newTodos)
+        saveToLocalStorage(newTodos)
+    }
+    const editTodo = (todoId: number, value: string) => {
+        const newTodos = todos.map(todoEl => {
+            if (todoId === todoEl.id) {
+                todoEl.title = value
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx]: do not mutate original objects in array

[docs](https://beta.reactjs.org/learn/updating-arrays-in-state#updating-objects-inside-arrays)

File: todo_react/todo-react/src/context/context.tsx [hunk 19]
```diff
@@ -0,0 +1,46 @@
+import React, {ChangeEvent, useContext} from "react";
+import {ITodo} from "../models";
+
+interface TodosContextModel {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/context.tsx:4]: you don't need ALL this data in context 😅
why `currentTitle` or `onEditChange` are available across whole app?
you should store in context only **common** data, which is needed in multiple components (todos list and methods to change them)

File: todo_react/todo-react/src/context/context.tsx [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    todos: ITodo[],
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/context.tsx:5]: `todos` and `todoList` are duplicates?

File: todo_react/todo-react/src/context/context.tsx [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    onEditChange: () => {}
+}
+
+const TodoContext = React.createContext<TodosContextModel>(TodosContextDefault);
+TodoContext.displayName = 'TodoContext';
+
+const useTodoContext = () => useContext(TodoContext);
+
+export {TodoContext, useTodoContext};
+export type {TodosContextModel};
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/context.tsx:46]: you can merge them in single export `export {TodosContextModel, TodoContext, useTodoContext};`

File: todo_react/todo-react/src/context/contextProvider.tsx [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+            let newEl = {...todoEl}
+            if (todoId === newEl.id) {
+                newEl.completed = !newEl.completed
+            }
+            return newEl
+        })
+        saveToDos(newTodos)
+    }
+    const editTodo = (todoId: number, value: string) => {
+        const newTodos = todos.map(todoEl => {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/context/contextProvider.tsx:47]: avoid redundant variables/copies

```js
const newTodos = todos.map(todoEl => {
  return todoEl.id === todoId
    ? {  ...todoEl, title: value }
    : todoEl
})
```

File: todo_react/todo-react/src/App.tsx [hunk 23]
```diff
@@ -0,0 +1,30 @@
+import React, {useEffect, useState} from 'react';
+import './App.css';
+import {AddToDo} from "./components/AddToDo";
+import {ToDoItem} from "./components/ToDoItem";
+import {useTodoContext} from "./context/context";
+
+export const App: React.FC = () => {
+    const todosContext = useTodoContext()
+    useEffect(() => {
```
Comment (gerasimvol) [inline todo_react/todo-react/src/App.tsx:9]: `<App>` is consumer of `todos` logic. Isn't better to manage all related logic (`useEffect` in this case) in `<TodoContextProvider>`?

File: todo_react/todo-react/src/components/AddToDo.tsx [hunk 24]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import {useTodoContext} from "../context/context";
+
+
+export const AddToDo: React.FC = () => {
+    const todosContext = useTodoContext()
+    return <div>
+        <form id="new-task-form">
+            <input type="text"
+                   value={todosContext.currentTitle}
```
Comment (gerasimvol) [inline todo_react/todo-react/src/components/AddToDo.tsx:10]: `currentTitle` is **local** state. why put in **global** context?

File: todo_react/todo-react/src/context/context.tsx (hunk 20 above)
Comment (KaterynaMadzhara) [reply todo_react/todo-react/src/context/context.tsx:5]: todoList - is the initial state for todos, I moved in a separate variable for easier readable

File: todo_react/todo-react/src/context/context.tsx (hunk 21 above)
Comment (KaterynaMadzhara) [reply todo_react/todo-react/src/context/context.tsx:46]: It did autocorrection, I will try

**Pull request merged**